sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala

```diff
@@ -62,8 +62,8 @@ class SQLMetric(
 
   override def merge(other: AccumulatorV2[Long, Long]): Unit = other match {
     case o: SQLMetric =>
-      if (o.isValid) {
-        if (!isValid) _value = defaultValidValue
+      if (!o.isZero) {
+        if (isZero) _value = 0
```

Comment:
```suggestion
        if (isZero) _value = defaultValidValue
```

Reply:
This was the issue before - in the case of an optional metric, `defaultValidValue` will be negative. So if we were to add a value to this metric, it would be subtracted by this sentinel value, so we want to set it to `0`.

We want `defaultValidValue` defines what is exposed via `value()`, not where to start counting - that's where I got confused the first time around.